### PR TITLE
feat: Add ability to use multiple signal servers

### PR DESCRIPTION
**Pull request by dmaretskyi**
- handle reconnect
- context to cancel events 
- support multiple concurent signal connections
- fix log

### CODE REVIEW

**Comment by mykola-vrmchk** [comment]
Current dependencies on/for this PR:
* main
  * **PR #2933** <a href="https://app.graphite.dev/github/pr/dxos/dxos/2933" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dxos/dxos/2933?utm_source=stack-comment).

File: packages/core/mesh/messaging/src/signal-client/signal-rpc-client.ts [hunk 1]
```diff
@@ -39,7 +50,15 @@ export class SignalRPCClient {
...
         send: (msg) => {
-          this._socket!.send(msg);
+          if (this._closed) {
+            // Do not send messages after close.
+            return;
+          }
+          try {
+            this._socket!.send(msg);
+          } catch (err) {
+            log.warn(String(err));
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-rpc-client.ts:60]:
```suggestion
            log.warn('send error', err);
```

File: packages/core/mesh/messaging/src/signal-client/signal-rpc-client.ts [hunk 2]
```diff
@@ -60,37 +79,53 @@ export class SignalRPCClient {
...
   }
 
   async close() {
+    this._closed = true;
     try {
       await this._rpc?.close();
+      this._socket?.close();
     } catch (err) {
-      log.catch(err);
+      log.warn(String(err));
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-rpc-client.ts:121]: log.warn('send error', err);

File: packages/core/mesh/messaging/src/signal-client/signal-client.ts [hunk 3]
```diff
@@ -295,47 +319,104 @@ export class SignalClient implements SignalMethods {
 
   private _reconnect() {
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-client.ts:320]: Turn this into DeferedTask to make sure it is only executed once at a time

File: packages/core/mesh/messaging/src/signal-client/signal-client.ts [hunk 4]
```diff
@@ -179,66 +220,86 @@ export class SignalClient implements SignalMethods {
...
     this._state = newState;
     this._lastStateChange = new Date();
     log('signal state changed', { status: this.getStatus() });
     this.statusChanged.emit(this.getStatus());
   }
 
-  private _initContext() {
-    this._ctx = new Context({
-      onError: (err) => log.catch(err)
+  private _initConnectionCtx() {
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-client.ts:253]: Merge into _createClient?

File: packages/core/mesh/messaging/src/signal-client/signal-client.ts [hunk 5]
```diff
@@ -179,66 +220,86 @@ export class SignalClient implements SignalMethods {
...
+          onConnected: () => {
+            log('socket connected');
+            this._lastError = undefined;
+            this._reconnectAfter = DEFAULT_RECONNECT_TIMEOUT;
+            this._setState(SignalState.CONNECTED);
+            this._clientReady.wake();
+            this._reconcileTask!.schedule();
+          },
+
+          onDisconnected: () => {
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-client.ts:282]: check if `onDisconnected` is always called even with errors and only call reconnect in here

Comment (dmaretskyi) [reply packages/core/mesh/messaging/src/signal-client/signal-client.ts:282]: add jsdoc for callbacks

File: packages/core/mesh/messaging/src/signal-client/signal-client.ts [hunk 6]
```diff
@@ -295,47 +319,104 @@ export class SignalClient implements SignalMethods {
...
+    }
+
+    // Subscribe to topics that are needed.
+    for (const { topic, peerId } of this._joinedTopics.values()) {
+      // Join desired topics.
+      if (this._swarmStreams.has({ topic, peerId })) {
+        continue;
+      }
+
+      const swarmStream = await client.join({ topic, peerId });
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-client.ts:372]: asyncTimeout(cancelWithContext(this._connectionCtx, <promise>), 5s)

File: packages/core/mesh/messaging/src/signal-client/signal-client.ts [hunk 7]
```diff
@@ -295,47 +319,104 @@ export class SignalClient implements SignalMethods {
...
+      this._messageStreams.delete(peerId);
+    }
+
+    // Subscribe to messages that are needed.
+    for (const { peerId } of this._subscribedMessages.values()) {
+      if (this._messageStreams.has(peerId)) {
+        continue;
+      }
+
+      const messageStream = await client.receiveMessages(peerId);
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-client.ts:408]: asyncTimeout(cancelWithContext(this._connectionCtx, ), 5s)

File: packages/core/mesh/messaging/src/signal-client/signal-client.ts [hunk 8]
```diff
@@ -179,66 +220,86 @@ export class SignalClient implements SignalMethods {
...
+          onDisconnected: () => {
+            log('socket disconnected', { state: this._state });
+            if (this._state !== SignalState.CONNECTED && this._state !== SignalState.CONNECTING) {
+              this._incrementReconnectTimeout();
+            }
+            this._setState(SignalState.DISCONNECTED);
+            this._reconnect();
+          },
+
+          onError: (error) => {
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-client.ts:291]: If onError and onDisconnect get called in-sequence does it create the client twice?

File: packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts [hunk 9]
```diff
@@ -67,30 +63,33 @@ export class WebsocketSignalManager implements SignalManager {
...
       return;
     }
+    log('open signal manager', { hosts: this._hosts });
     log.trace('dxos.mesh.websocket-signal-manager', trace.begin({ id: this._instanceId, parentId: this._traceParent }));
 
     this._initContext();
 
     [...this._servers.values()].forEach((server) => server.open());
 
+    await Promise.all(
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts:81]: await this._forEachServer(async server => {
  ...
})

File: packages/core/mesh/messaging/src/signal-client/signal-client.ts [hunk 10]
```diff
@@ -179,66 +220,86 @@ export class SignalClient implements SignalMethods {
...
-
-    this._ctx.onDispose(() => {
-      messageStream.close();
-      this._messageStreams.delete(peerId);
-    });
+    log('subscribing to messages', { peerId });
+    this._subscribedMessages.add({ peerId });
+    this._reconcileTask!.schedule();
 
     return {
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-client.ts:227]: change to a separate this.unsubscibeMessages method

File: packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts [hunk 11]
```diff
@@ -23,17 +23,14 @@ export class WebsocketSignalManager implements SignalManager {
   private readonly _servers = new Map<string, SignalClient>();
 
   /** Topics joined: topic => peerId */
-  private readonly _topicsJoined = new ComplexMap<PublicKey, PublicKey>((topic) => topic.toHex());
-
-  /** host => topic => peerId */
-  private readonly _topicsJoinedPerSignal = new Map<string, ComplexMap<PublicKey, PublicKey>>();
+  private readonly _topicsJoined = new ComplexSet<{ topic: PublicKey; peerId: PublicKey }>(
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts:26]: see if the state in individual signal clients is enough

File: packages/core/mesh/messaging/src/signal-client/signal-rpc-client.ts [hunk 12]
```diff
@@ -17,19 +17,30 @@ interface Services {
...
+  onDisconnected?: () => void;
+  onError?: (error: Error) => void;
+};
+
+export type SignalRPCClientParams = {
+  url: string;
+  callbacks?: SignalCallbacks;
+};
+
 export class SignalRPCClient {
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-client/signal-rpc-client.ts:31]: use genric websocket dxrpc client class

**Pull request merged**